### ISSUE TITLE
Defer synchronous reception report with out-of-order data segments

### ISSUE THREAD

**Issue by BrianSipos** [nasa/HDTN#22]
When red part data is segmented and delivered to the receiving engine out-of-order, the checkpoint(s) and EORP can be received before the earlier-in-block data segments. If a synchronous report is sent immediately upon receiving the checkpoint there will be data segments in-flight and about to be delivered that will be seen as reception gaps in the report.

Instead of sending the synchronous report immediately upon receiving a checkpoint segment the receiving engine should have some more complex logic:
 1. If the report segment bounds are fully claimed (i.e. no gaps) then the report can be sent immediately.
 2. Otherwise, the engine should wait a very small window of time for gaps to be filled. The delay time should reset upon any data segments which fill gaps.

In a situation with no loss but lots of out-of-order delivery this will have exactly the same number of reports, they will just be sent when the full checkpointed bounds of data have been received. In a situation with loss this will send reports with the fewest size of claim gaps.

**Comment by briantomko** [comment]
@BrianSipos just an update, I've implemented this but still working on testing it.  Will let you know.

**Comment by briantomko** [comment]
@BrianSipos I got this on side branch 111-defer-reception-reports if you would like to test.  It only resolves issue 22 right now.  My unit test in TestLtpUdpEngine.cpp, function DoTestReverseStartToEOBSrcToDest reverses the packet order.  The last parameter of AddLtpUdpEngine lets you specify the time delay in milliseconds.  On clean exit, the LtpEngine destructor will print the following five additional stats:
    uint64_t m_numGapsFilledByOutOfOrderDataSegments;
    uint64_t m_numDelayedFullyClaimedPrimaryReportSegmentsSent;
    uint64_t m_numDelayedFullyClaimedSecondaryReportSegmentsSent;
    uint64_t m_numDelayedPartiallyClaimedPrimaryReportSegmentsSent;
    uint64_t m_numDelayedPartiallyClaimedSecondaryReportSegmentsSent;

Also, while revisiting this code, I found and fixed several performance issues like unnecessary std::map and map operations, so it may run faster.
